lib/rules/no-misleading-character-class.js

```diff
@@ -371,12 +362,19 @@ module.exports = {
                 for (const { node: refNode } of tracker.iterateGlobalReferences({
                     RegExp: { [CALL]: true, [CONSTRUCT]: true }
                 })) {
+                    let pattern, flags;
                     const [patternNode, flagsNode] = refNode.arguments;
-                    const pattern = getStringIfConstant(patternNode, scope);
-                    const flags = getStringIfConstant(flagsNode, scope);
 
-                    if (typeof pattern === "string") {
-                        verify(patternNode, pattern, flags || "", fixer => {
+                    if (patternNode.type === "Literal" && patternNode.regex) {
+                        pattern = patternNode.regex.pattern;
+                        flags = flagsNode ? getStringIfConstant(flagsNode, scope) : patternNode.regex.flags;
```

Comment:
If there's no `flagsNode`, perhaps we could just skip checking the regex here because it has already been verified as a literal?

![image](https://github.com/eslint/eslint/assets/44349756/5aef8907-8b9d-4bdb-9161-224734f4b27d)

Reply:
Makes sense 👍🏻 Shall we also try to avoid duplicate reports when there is a `flags` argument alongside a regex literal? For example:

![image](https://github.com/eslint/eslint/assets/6367844/93157ecc-2a9f-4014-be82-3cbf6c110d15)

That could get tricky because the reports don't need to be granular, so we'd have to compare the character matches.

Reply:
> Shall we also try to avoid duplicate reports when there is a `flags` argument alongside a regex literal?

Yes, I think it would be good to avoid duplicate reports. Perhaps in that case the rule should skip checking the regex literal and check only the constructor call?

For instance, the following might be considered a false positive because the regex created by the literal isn't actually used as such?

```js
/*eslint no-misleading-character-class: error */

RegExp(/^[👍]$/, "u"); // error on the regex literal
```

> That could get tricky because the reports don't need to be granular, so we'd have to compare the character matches.

I'm not sure I understand this, can you clarify? If the source of the pattern is a regex literal, than can't we just use regexpp's indexes?

Reply:
> Yes, I think it would be good to avoid duplicate reports. Perhaps in that case the rule should skip checking the regex literal and check only the constructor call?

Yes, we could do that if the argument is a regex literal. If it's a variable, it could be used in other places.

> > That could get tricky because the reports don't need to be granular, so we'd have to compare the character matches.
> 
> I'm not sure I understand this, can you clarify? If the source of the pattern is a regex literal, than can't we just use regexpp's indexes?

In afterthought, I'd say it would be better not to check regex literals that are supplied as arguments to the `RegExp` constructor like you suggested. At least I think this is what should be done if the `RegExp` call has a `flags` argument. If a regex is passed as a variable, it should be fine to keep the duplicate reports.

So the whole story would look like this:

```js
/*eslint no-misleading-character-class: error */

RegExp(/[👍]/);                     // error only on the regex literal

const regex = /[👍]/;               // error
RegExp(regex);                      // no error

RegExp(/[👍]/, "");                 // error only on the RegExp call

const regex1 = /[👍]/g;             // error
doSomethingElse(regex1);
const regex2 = RegExp(regex1, "g"); // also error
```

Reply:
Sounds good to me.

So, if the `RegExp` constructor is called with one argument that is a regex value, we'll skip checking the constructor call. If regex literal is the first argument of `RegExp` constructor call that also has the second argument, we'll skip checking the regex literal.

Reply:
Updated in 6792594b0423363b2d423faa35c2ce45d444156a.